apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,502 @@
+
+        // TODO: Should this be created with a MetadataClient instead?
+        private static readonly ILogger Logger = ApplicationContext.Logger.ForType<MetadataClientImpl>();
+
+        private readonly string metadataHost;
+
+        private readonly object eventLock = new object();
+        private ValueWatcher instanceWatcher;
+        private ValueWatcher maintenanceStatusWatcher;
+        private ValueWatcher projectWatcher;
+
+        public override ConfigurableHttpClient HttpClient { get; }
+
+        /// <summary>
+        /// Constructs a new client for the metadata server.
+        /// </summary>
+        /// <param name="httpClient">The <see cref="ConfigurableHttpClient"/> to use when making requests, or null to create a default instance.</param>
+        public MetadataClientImpl(ConfigurableHttpClient httpClient = null)
+        {
+            HttpClient = httpClient ?? new HttpClientFactory().CreateHttpClient(new CreateHttpClientArgs());
+            metadataHost = GetMetadataHost();
+        }
+
+        /// <summary>
+        /// Constructs a new client for the metadata server.
+        /// </summary>
+        /// <param name="httpClientFactory">
+        /// An <see cref="IHttpClientFactory"/> used to create a <see cref="ConfigurableHttpClient"/> for making requests, or null to
+        /// create a default instance.
+        /// </param>
+        public MetadataClientImpl(IHttpClientFactory httpClientFactory)
+            : this((httpClientFactory ?? new HttpClientFactory()).CreateHttpClient(new CreateHttpClientArgs())) { }
+
+        private static MaintenanceStatus GetMaintenanceStatus(string result)
+        {
+            switch (result)
+            {
+                case "NONE": return MaintenanceStatus.None;
+                case "MIGRATE_ON_HOST_MAINTENANCE": return MaintenanceStatus.Migrate;
+                default: return MaintenanceStatus.Unknown;
+            }
+        }
+
+        private string GetMetadataHost()
+        {
+            var emulatorHost = Environment.GetEnvironmentVariable(EmulatorEnvironmentVariable);
+            return string.IsNullOrEmpty(emulatorHost) ? DefaultMetadataHost : emulatorHost;
+        }
+
+        private static T GetResult<T>(Func<CancellationToken, Task<T>> operation)
+        {
+            try
+            {
+                return operation(CancellationToken.None).Result;
+            }
+            catch (AggregateException e)
+            {
+                throw e.InnerExceptions.FirstOrDefault() ?? e;
+            }
+        }
+
+        private async Task<HttpResponseMessage> RequestMetadata(string relativeUrl, CancellationToken cancellationToken)
+        {
+            var httpRequest = new HttpRequestMessage(HttpMethod.Get, $"http://{metadataHost}/computeMetadata/v1/{relativeUrl}");
+            httpRequest.Headers.Add(MetadataFlavor, GoogleMetadataHeader);
+            return await HttpClient.SendAsync(httpRequest, cancellationToken).ConfigureAwait(false);
```

Comment:
Don't really need to await the task here, you can just return it. Unless you want to handle the error here of course.